ReactWindows/ReactNative.Shared/UIManager/BorderedCanvas.cs

```diff
@@ -47,6 +52,298 @@ public Border Border
+        /// </summary>
+        protected bool HasBorder()
+        {
+            return Border != null;
+        }
+
+        /// <summary>
+        /// Adds a Border to a Canvas if it hasn't been added already.
+        /// </summary>
+        protected Border GetOrCreateBorder()
+        {
+            if (Border == null)
+            {
+                var borderProps = GetBorderProps();
+                var borderBrush = (borderProps?.Color.HasValue ?? false) ? new SolidColorBrush(borderProps.Color.Value) : DefaultBorderBrush;
+                Border = new Border { BorderBrush = borderBrush };
+
+                // Layout animations bypass SetDimensions, hence using XAML bindings.
+
+                Border.SetBinding(FrameworkElement.WidthProperty, new Binding
+                {
+                    Source = this,
+                    Path = new PropertyPath("Width")
+                });
+
+                Border.SetBinding(FrameworkElement.HeightProperty, new Binding
+                {
+                    Source = this,
+                    Path = new PropertyPath("Height")
+                });
+            }
+
+            return Border;
+        }
+
+        /// <summary>
+        /// Gets or sets background brush of Canvas itself or of Border child as
+        /// appropriate. This is the only supported method to access background
+        /// of BorderedCanvas. DO NOT use Canvas.Background or Border.Background directly.
+        /// </summary>
+        public new Brush Background
+        {
+            get
+            {
+                if (Border == null)
+                {
+                    return base.Background;
+                }
+                else
+                {
+                    var border = GetOrCreateBorder();
+                    return border.Background;
+                }
+            }
+            set
+            {
+                if (Border == null)
+                {
+                    base.Background = value;
+                }
+                else
+                {
+                    var border = GetOrCreateBorder();
+                    border.Background = value;
+                    base.Background = base.Background ?? _defaultBackgroundBrush;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the border radius of the view.
+        /// </summary>
+        public CornerRadius BorderRadius
+        {
+            get
+            {
+                if (HasBorder())
+                {
+                    var border = GetOrCreateBorder();
+                    return border.CornerRadius;
+                }
+                else
+                {
+                    return new CornerRadius();
+                }
+            }
+            set
+            {
+                var border = GetOrCreateBorder();
+                border.CornerRadius = value;
+                TransferBackgroundBrush();
```

Comment:
>TransferBackgroundBrush [](start = 16, length = 23)

Why do we call this here rather than in `GetOrCreateBorder`?

It seems like the code in `TransferBackgroundBrush` has to run just once when we create the border.

We could achieve that by putting the `TransferBackgroundBrush` logic into the branch of `GetOrCreateBorder` that creates the border.

Instead of doing that, it looks like we have the caller of `GetOrCreateBorder` call `TransferBackgroundBrush` whenever there's a chance `GetOrCreateBorder` might have created a border.

Thoughts?

Reply:
I agree the original code reached the stage where it would be good to refactor it. But I have 3 arguments not to do this in this PR:

1. This PR is to enable the gradient implemnentation in a parallel PR. Doing refactoring like this is not needed for gradient support and should be done as separate change.

2. I saw some talks on other RNW PRs about possible upcoming changes in this part of code related to adding more general support for optional children (shadows and such) that would probably introduce different mechanisms for optional children management and so to the question of GetOrCreateBorder/TransferBackgrounddBrush - for example possibility to remove the optional children. So the work of doing this refactoriong would be lost anyway.

3. The refactoring would probably need to be deeper since currently GetOrCreateBorder is used also for cases when we already know border exists, so ideally we should not call Transfer in those cases. In general current combination of Border+HasBorder+GetOrCreateBorder private API is not ideal and should be redone as a whole. But this will make argument 2 even more prominent.

Do you think we should create a separate work item for the refactoring, or do you still think we should do this refactoring in this PR?

---
In reply to: [177619360](https://github.com/Microsoft/react-native-windows/pull/1730#discussion_r177619360) [](ancestors = 177619360)

Reply:
Doing it in a separate PR sounds fine to me.

> The refactoring would probably need to be deeper since currently GetOrCreateBorder is used also for cases when we already know border exists, so ideally we should not call Transfer in those cases.

Isn't that easy to achieve? We put the call to `TransferBackgrounddBrush` inside of the branch of `GetOrCreateBorder` that creates the border.